### PR TITLE
Centernet

### PR DESCRIPTION
**Pull request by zhreshold**
Add model + training pipeline for centernet: https://arxiv.org/abs/1904.07850?context=cs

DLA(deep layer aggregation model series are still under development), resnet variants are ready and outperformed paper results.

I will include pretrained models in a separate PR.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1029-17 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1029/17/index.html 
Code coverage of this PR:  vs. Master:

**Comment by hetong007** [comment]
Please let us know when it is ready to be reviewed.

**Comment by zhreshold** [comment]
@hetong007 @bryanyzhu @Jerryzcn @xinyu-intel Ready to be reviewed, the website contents needs to be added in a separate PR since this one is already huge.

File: gluoncv/model_zoo/center_net/deconv_dla.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            ida = self.idas[i]
+            ida.startp = len(layers) -i - 2
+            ida.endp = len(layers)
+            layers = ida(layers)
+            out.insert(0, layers[-1])
+        return out
+
+
+class DLASeg(nn.HybridBlock):
+    """Short summary.
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/center_net/deconv_dla.py]: Can you provide more details for DLASeg?

File: gluoncv/data/transforms/bbox.py [hunk 2]
```diff
@@ -160,3 +160,125 @@ def translate(bbox, x_offset=0, y_offset=0):
...
+    -------
+    numpy.ndarray
+        New bounding box with shape (1, 2).
+
+    """
+    new_pt = np.array([pt[0], pt[1], 1.], dtype=np.float32).T
+    new_pt = np.dot(t, new_pt)
+    return new_pt[:2]
+
+def get_dir(src_point, rot_rad):
```
Comment (bryanyzhu) [inline gluoncv/data/transforms/bbox.py]: IMO, 'get_dir' is easily confused with get directory. Maybe using 'get_direction' or 'get_rot_dir' is more clear.

File: gluoncv/model_zoo/dla.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    expansion = 2
+    cardinality = 32
+
+    def __init__(self, inplanes, planes, stride=1, dilation=1,
+                 norm_layer=BatchNorm, norm_kwargs=None, **kwargs):
+        super(BottleneckX, self).__init__(**kwargs)
+        if norm_kwargs is None:
+            norm_kwargs = {}
+        cardinality = BottleneckX.cardinality
+        # dim = int(math.floor(planes * (BottleneckV5.expansion / 64.0)))
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/dla.py]: Please remove comments if unnecessary.

File: scripts/detection/center_net/train_center_net.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        LRScheduler(args.lr_mode, base_lr=args.lr,
+                    nepochs=args.epochs - args.warmup_epochs,
+                    iters_per_epoch=num_batches,
+                    step_epoch=lr_decay_epoch,
+                    step_factor=args.lr_decay, power=2),
+    ])
+
+    for k, v in net.collect_params('.*bias').items():
+        v.wd_mult = 0.0
+    # trainer = gluon.Trainer(
```
Comment (bryanyzhu) [inline scripts/detection/center_net/train_center_net.py]: Please remove comments if unnecessary. And is adam better than sgd?

**Comment by mli** [comment]
Job PR-1029-18 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1029/18/index.html 
Code coverage of this PR:  vs. Master:

File: scripts/detection/center_net/train_center_net.py (hunk 4 above)
Comment (zhreshold) [reply scripts/detection/center_net/train_center_net.py]: similar, though I don't have fully compariable results in the latest version

File: gluoncv/model_zoo/dla.py (hunk 3 above)
Comment (zhreshold) [reply gluoncv/model_zoo/dla.py]: done

File: gluoncv/data/transforms/bbox.py (hunk 2 above)
Comment (zhreshold) [reply gluoncv/data/transforms/bbox.py]: done

File: gluoncv/model_zoo/center_net/deconv_dla.py (hunk 1 above)
Comment (zhreshold) [reply gluoncv/model_zoo/center_net/deconv_dla.py]: done

**Comment by mli** [comment]
Job PR-1029-20 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1029/20/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/data/transforms/presets/center_net.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        self._width = width
+        self._height = height
+        self._num_class = num_class
+        self._scale_factor = scale_factor
+        self._mean = np.array(mean, dtype=np.float32).reshape(1, 1, 3)
+        self._std = np.array(std, dtype=np.float32).reshape(1, 1, 3)
+        self._data_rng = np.random.RandomState(123)
+        self._eig_val = np.array([0.2141788, 0.01817699, 0.00341571],
+                                 dtype=np.float32)
+        self._eig_vec = np.array([
```
Comment (hetong007) [inline gluoncv/data/transforms/presets/center_net.py]: where can I find reference of this matrix? As it is similar but not identical with https://github.com/apache/incubator-mxnet/blob/master/src/io/image_aug_default.cc#L545

File: gluoncv/model_zoo/center_net/deconv_resnet.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                       net.maxpool,
+                       net.layer1,
+                       net.layer2,
+                       net.layer3,
+                       net.layer4])
+            self.base_network = feat
+        else:
+            raise NotImplementedError()
+        with self.name_scope():
+            self.deconv = self._make_deconv_layer(deconv_filters, deconv_kernels)
```
Comment (hetong007) [inline gluoncv/model_zoo/center_net/deconv_resnet.py]: in pose estimation pixel shuffle achieves similar acc with much less computation comparing to deconv, will we consider that?

File: gluoncv/data/transforms/presets/center_net.py (hunk 5 above)
Comment (zhreshold) [reply gluoncv/data/transforms/presets/center_net.py]: It's origin is hard to track now, maybe from Corner implementation. The values are calculated using COCO dataset statistics but I guess it's rather subtle for training on difference dataset. I can take a look and see if the eigenvalues/eigenvector are important in the future

File: gluoncv/model_zoo/center_net/deconv_resnet.py (hunk 6 above)
Comment (zhreshold) [reply gluoncv/model_zoo/center_net/deconv_resnet.py]: There's a plan for developing a fast version so pixel shuffle could be a better choice
